service/worker/sysworkflow/system_workflow.go

```diff
@@ -22,101 +22,337 @@ package sysworkflow
+	globalMetricsClient.IncCounter(metrics.SystemWorkflowScope, metrics.SysWorkerContinueAsNew)
+	sw.Stop()
+	logger.WithFields(bark.Fields{
+		logging.TagNumberOfSignalsUntilContinueAsNew: signalsHandled,
+	}).Info("system workflow is continuing as new")
+	return workflow.NewContinueAsNewError(ctx, systemWorkflowFnName)
 }
 
-func selectSystemTask(scope tally.Scope, signal signal, ctx workflow.Context, logger *zap.Logger) {
-	scope.Counter(HandledSignalCount).Inc(1)
-
+func selectSystemTask(signal signal, ctx workflow.Context, logger bark.Logger, metricsClient metrics.Client) {
 	ao := workflow.ActivityOptions{
 		ScheduleToStartTimeout: time.Minute,
 		StartToCloseTimeout:    time.Minute,
 		HeartbeatTimeout:       time.Second * 10,
 		RetryPolicy: &cadence.RetryPolicy{
-			InitialInterval:          time.Second,
-			BackoffCoefficient:       2.0,
-			MaximumInterval:          time.Minute,
-			ExpirationInterval:       time.Hour * 24 * 30,
-			MaximumAttempts:          0,
-			NonRetriableErrorReasons: []string{},
+			InitialInterval:    time.Second,
+			BackoffCoefficient: 2.0,
+			MaximumInterval:    time.Minute,
+			ExpirationInterval: time.Hour * 24 * 30,
+			NonRetriableErrorReasons: []string{
+				errArchivalUploadActivityGetDomainStr,
+				errArchivalUploadActivityNextBlobStr,
+				errArchivalUploadActivityConstructKeyStr,
+				errArchivalUploadActivityBlobExistsStr,
+				errArchivalUploadActivityMarshalBlobStr,
+				errArchivalUploadActivityConvertHeaderToTagsStr,
+				errArchivalUploadActivityWrapBlobStr,
+				errArchivalUploadActivityUploadBlobStr,
+				errDeleteHistoryActivityDeleteFromV2Str,
+				errDeleteHistoryActivityDeleteFromV1Str,
+			},
 		},
 	}
 
 	actCtx := workflow.WithActivityOptions(ctx, ao)
 	switch signal.RequestType {
 	case archivalRequest:
+		request := *signal.ArchiveRequest
 		if err := workflow.ExecuteActivity(
 			actCtx,
-			ArchivalActivityFnName,
-			*signal.ArchiveRequest,
+			archivalUploadActivityFnName,
+			request,
 		).Get(ctx, nil); err != nil {
-			scope.Counter(ArchivalFailureErr)
-			logger.Error("failed to execute archival activity", zap.Error(err))
+			logger.WithFields(bark.Fields{
+				logging.TagErr:                             err,
+				logging.TagArchiveRequestDomainID:          request.DomainID,
+				logging.TagArchiveRequestWorkflowID:        request.WorkflowID,
+				logging.TagArchiveRequestRunID:             request.RunID,
+				logging.TagArchiveRequestEventStoreVersion: request.EventStoreVersion,
+				logging.TagArchiveRequestLastFirstEventID:  request.LastFirstEventID,
+			}).Error("ArchivalUploadActivity encountered non-retryable error")
+			metricsClient.IncCounter(metrics.SystemWorkflowScope, metrics.SysWorkerArchivalUploadActivityNonRetryableFailures)
+		} else {
+			metricsClient.IncCounter(metrics.SystemWorkflowScope, metrics.SysWorkerArchivalUploadSuccessful)
+		}
+		if err := workflow.ExecuteActivity(
```

Comment:
Delete history is usually one API call to persistence layer.  Can this be converted to Local Activity.

Reply:
Converted it to local activity. I assume that as long as I do retries of transient errors in the activity then this will be guaranteed to delete history still. Lets chat in person about this assumption.